Sources/StreamVideoSwiftUI/Utils/PictureInPicture/StreamPictureInPictureVideoRenderer.swift

```diff
@@ -0,0 +1,134 @@
+        super.init(frame: frame)
+        setUp()
+    }
+
+    override func willMove(toWindow newWindow: UIWindow?) {
+        if newWindow != nil {
+            startPictureInPicture(for: track, on: newWindow)
+        } else {
+            stopPictureInPicture(for: track)
+        }
+        super.willMove(toWindow: newWindow)
+    }
+
+    // MARK: - Rendering lifecycle
+
+    func setSize(_ size: CGSize) {
+        scaleFactor = size.height > size.width
+        ? size.height / size.width
+        : size.width / size.height
+    }
+
+    func renderFrame(_ frame: RTCVideoFrame?) {
+        guard let frame = frame else {
+            return
+        }
+
+        if let pixelBuffer = frame.buffer as? RTCCVPixelBuffer {
+            guard let sampleBuffer = CMSampleBuffer.from(pixelBuffer.pixelBuffer) else {
+                log.warning("Failed to convert CVPixelBuffer to CMSampleBuffer")
+                return
+            }
+
+            bufferPublisher.send(sampleBuffer)
+        } else if let i420buffer = frame.buffer as? RTCI420Buffer {
+            // We reduce the track resolution, since it's displayed in a smaller place.
+            // Values are picked depending on how much the PiP view takes in an average iPhone or iPad.
+            let reductionFactor = UIDevice.current.isIpad ? 4 : 6
+            guard let buffer = convertI420BufferToPixelBuffer(i420buffer, reductionFactor: reductionFactor),
+                  let sampleBuffer = CMSampleBuffer.from(buffer) else {
+                return
+            }
+
+            bufferPublisher.send(sampleBuffer)
+        }
+    }
+
+    // MARK: - Private helpers
+
+    private func setUp() {
+        addSubview(contentView)
+        NSLayoutConstraint.activate([
+            contentView.leadingAnchor.constraint(equalTo: leadingAnchor),
+            contentView.trailingAnchor.constraint(equalTo: trailingAnchor),
+            contentView.topAnchor.constraint(equalTo: topAnchor),
+            contentView.bottomAnchor.constraint(equalTo: bottomAnchor)
+        ])
+    }
+
+    private func process(_ buffer: CMSampleBuffer) {
+        guard let trackId = track?.trackId else {
+            contentView.sampleBufferDisplayLayer.flush()
+            return
+        }
+
+        if #available(iOS 14.0, *) {
+            if contentView.sampleBufferDisplayLayer.requiresFlushToResumeDecoding == true {
+                contentView.sampleBufferDisplayLayer.flush()
+                log.debug("→ Display layer for track:\(trackId) flushed ✅")
+            }
+        }
+        if contentView.sampleBufferDisplayLayer.isReadyForMoreMediaData == true {
+            contentView.sampleBufferDisplayLayer.enqueue(buffer)
+//            log.debug("→ Frame for track:\(trackId) enqueued on picture in picture buffer ✅")
+        }
+    }
+
+    private func startPictureInPicture(
+        for track: RTCVideoTrack?,
+        on window: UIWindow?
+    ) {
+        guard window != nil, let track else { return }
```

Comment:
Why do we need the window, could you explain?

Reply:
I have added comments now explaining those files, but to reiterate here, if the Picture in Picture window isn't visible, then we aren't consuming video frames, resulting in saving some computing operations.